File: solutions/fable-1/fable/zig/brackets.zig
const std = @import("std");

fn readAllStdin(allocator: std.mem.Allocator) ![]u8 {
    const stdin = std.fs.File.stdin();
    var buf = try allocator.alloc(u8, 1 << 16);
    var len: usize = 0;
    while (true) {
        if (len == buf.len) {
            buf = try allocator.realloc(buf, buf.len * 2);
        }
        const n = try stdin.read(buf[len..]);
        if (n == 0) break;
        len += n;
    }
    return buf[0..len];
}

fn balanced(line: []const u8, stack: []u8) bool {
    var sp: usize = 0;
    for (line) |c| {
        switch (c) {
            '(', '[', '{' => {
                stack[sp] = c;
                sp += 1;
            },
            ')' => {
                if (sp == 0 or stack[sp - 1] != '(') return false;
                sp -= 1;
            },
            ']' => {
                if (sp == 0 or stack[sp - 1] != '[') return false;
                sp -= 1;
            },
            '}' => {
                if (sp == 0 or stack[sp - 1] != '{') return false;
                sp -= 1;
            },
            else => {},
        }
    }
    return sp == 0;
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const input = try readAllStdin(allocator);
    const stack = try allocator.alloc(u8, input.len + 1);

    var out = try allocator.alloc(u8, input.len + 16);
    var olen: usize = 0;

    var it = std.mem.splitScalar(u8, input, '\n');
    var first = true;
    while (it.next()) |raw| {
        // Skip the empty tail after a final newline (but keep a leading-only empty input? no lines then).
        if (it.peek() == null and raw.len == 0) {
            if (first) break; // empty input: no lines
            break;
        }
        first = false;
        var line = raw;
        if (line.len > 0 and line[line.len - 1] == '\r') line = line[0 .. line.len - 1];
        const ans: []const u8 = if (balanced(line, stack)) "yes\n" else "no\n";
        if (olen + ans.len > out.len) {
            out = try allocator.realloc(out, out.len * 2 + ans.len);
        }
        @memcpy(out[olen .. olen + ans.len], ans);
        olen += ans.len;
    }

    try std.fs.File.stdout().writeAll(out[0..olen]);
}

File: solutions/fable-1/fable/rust/brackets.rs
use std::io::{self, BufRead, Write};

fn balanced(line: &str) -> bool {
    let mut stack: Vec<u8> = Vec::new();
    for b in line.bytes() {
        match b {
            b'(' | b'[' | b'{' => stack.push(b),
            b')' | b']' | b'}' => {
                let expected = match b {
                    b')' => b'(',
                    b']' => b'[',
                    _ => b'{',
                };
                if stack.pop() != Some(expected) {
                    return false;
                }
            }
            b'\r' | b' ' | b'\t' => {}
            _ => return false,
        }
    }
    stack.is_empty()
}

fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut out = io::BufWriter::new(stdout.lock());
    for line in stdin.lock().lines() {
        let line = match line {
            Ok(l) => l,
            Err(_) => break,
        };
        let ans = if balanced(&line) { "yes" } else { "no" };
        writeln!(out, "{}", ans).unwrap();
    }
    out.flush().unwrap();
}
